crates/circuit/src/dag_circuit.rs

```diff
@@ -1762,15 +1762,13 @@ def _format(operand):
             let clbits_id = self
                 .cargs_interner
                 .insert_owned(self.clbits.map_bits(cargs.iter().flatten())?.collect());
-            let instr = PackedInstruction {
-                op: py_op.operation,
-                qubits: qubits_id,
-                clbits: clbits_id,
-                params: (!py_op.params.is_empty()).then(|| Box::new(py_op.params)),
-                extra_attrs: py_op.extra_attrs,
-                #[cfg(feature = "cache_pygates")]
-                py_op: op.unbind().into(),
```

Comment:
Yes, but also think of it this way. 

If you look at the `PackedInstruction::unpack_py_op()` method, in which a reference is built to a python object (basically what we do when the `cache_pygate` flag is disabled), in the case of a standard gate, we build a new instance of py_op based on the current parameters or the extra attributes of our gate, and not only that, but the instance we get from this function whenever dealing with any other type of gate comes directly from the `OperationRef` type if it contains a python object within and isn't really affected by changes unless our `op` type is `StandardGate`.

Why do I mention this? Well, whenever a gate is initialized without a cached pygate, this is the function in charge of building that gate for us. Since this gate is disconnected and changes won't propagate, we'd need to update it by creating a new one whenever our params or attributes are modified (or at least that's what I understand from here).

We could technically optimize further by only re-building if we have an instance of `Standard` gate, since what we do in other cases is a call to `clone_ref(py). What do you think?

Reply:
Python doesn't enforce ownership in the way Rust does, but we can still think of it in similar terms. `DAGCircuit.apply_operation_back` has always behaved _as if_ calling it transferred ownership of the gate object, as opposed to `QuantumCircuit.append`, which copied the object if it detected that you might want to modify it (i.e. it was parametric).

Since the Python object coming in is logically "owned" by the function now, I think it's fine to put it in the cache.

With all the Python stuff, we can't compile-time (or even, in most cases, run-time) forbid modifications to Python objects, so there's always going to be the possibility for a user to break us a bit, and some of our Python-based will be pragmatics.  I think in this case, saving the incoming object in the cache is no more dangerous than having the cache at all, so it should be fine to keep.